From 4f750816403af1bc218b7800c553f58ec4188244 Mon Sep 17 00:00:00 2001
From: Dinesh Sudham <dineshx.sudham@intel.com>
Date: Wed, 4 Nov 2020 17:00:28 +0800
Subject: [PATCH] DRVLIB_SW-3941:GSWAPI: Marking TFLOW rule entry is valid or
 invalid

PCE RULE can be mark as invalid without deleting Rule pattern and action,
this will enable the mark rule valid or invalid without rule deletion.

Signed-off-by: Dinesh Sudham <dineshx.sudham@intel.com>
---
 .../net/ethernet/lantiq/switch-api/gsw_flow_core.c | 74 ++++++++++++++++++++++
 .../net/ethernet/lantiq/switch-api/gsw_ll_func.h   |  3 +
 .../net/ethernet/lantiq/switch-api/gsw_tbl_rw.c    | 50 +++++++++++++++
 include/net/switch_api/gsw_flow_ops.h              |  6 +-
 include/net/switch_api/gsw_tbl_rw.h                |  1 +
 include/net/switch_api/lantiq_gsw_flow.h           | 26 ++++++++
 6 files changed, 159 insertions(+), 1 deletion(-)

diff --git a/drivers/net/ethernet/lantiq/switch-api/gsw_flow_core.c b/drivers/net/ethernet/lantiq/switch-api/gsw_flow_core.c
index a29dac8df592..251aa0621678 100644
--- a/drivers/net/ethernet/lantiq/switch-api/gsw_flow_core.c
+++ b/drivers/net/ethernet/lantiq/switch-api/gsw_flow_core.c
@@ -27880,6 +27880,78 @@ GSW_return_t GSW_PceRuleFree(void *cdev, GSW_PCE_rule_alloc_t *parm)
 	return ret;
 }
 
+/**
+ * GSW_PceRuleEnable()- Enable PCE Rule
+ * @cdev: a pointer to the target &ethsw_api_dev_t
+ * @GSW_PCE_rule_t: a pointer to PCE Rule target &GSW_PCE_rule_t
+ *
+ * Parse the valid bit,Table Address and Enable PCE Rule.
+ */
+GSW_return_t GSW_PceRuleEnable(void *cdev,
+			       GSW_PCE_rule_t *parm)
+{
+	ethsw_api_dev_t *gswdev = GSW_PDATA_GET(cdev);
+	pctbl_prog_t pcetable;
+	int ret;
+
+	BUG_ON(!cdev || !gswdev);
+
+	if (GSW_statusOk != gsw_pcerulemanage(cdev, parm))
+		return GSW_statusErr;
+
+	memset(&pcetable, 0, sizeof(pctbl_prog_t));
+	pcetable.table = PCE_TFLOW_INDEX;
+	pcetable.pcindex = parm->pattern.nIndex;
+	gsw_pce_table_read(cdev, &pcetable);
+	memset(&pcetable, 0, sizeof(pctbl_prog_t));
+	pcetable.pcindex = parm->pattern.nIndex;
+	pcetable.valid = true;
+	pcetable.table = PCE_TFLOW_INDEX;
+	ret = gsw_pce_rule_set(cdev, &pcetable);
+	if (ret) {
+		pr_err("%s:%s:%d", __FILE__, __func__, __LINE__);
+		return GSW_statusErr;
+	}
+
+	return GSW_statusOk;
+}
+
+/**
+ * GSW_PceRuleDisable()- Disable PCE Rule
+ * @cdev: a pointer to the target &ethsw_api_dev_t
+ * @GSW_PCE_rule_t: a pointer to PCE Rule target &GSW_PCE_rule_t
+ *
+ * Parse the valid bit,Table Address and Disable PCE Rule.
+ */
+GSW_return_t GSW_PceRuleDisable(void *cdev,
+				GSW_PCE_rule_t *parm)
+{
+	ethsw_api_dev_t *gswdev = GSW_PDATA_GET(cdev);
+	pctbl_prog_t pcetable;
+	int ret;
+
+	BUG_ON(!cdev || !gswdev);
+
+	if (GSW_statusOk != gsw_pcerulemanage(cdev, parm))
+		return GSW_statusErr;
+
+	memset(&pcetable, 0, sizeof(pctbl_prog_t));
+	pcetable.table = PCE_TFLOW_INDEX;
+	pcetable.pcindex = parm->pattern.nIndex;
+	gsw_pce_table_read(cdev, &pcetable);
+	memset(&pcetable, 0, sizeof(pctbl_prog_t));
+	pcetable.pcindex = parm->pattern.nIndex;
+	pcetable.valid = false;
+	pcetable.table = PCE_TFLOW_INDEX;
+	ret = gsw_pce_rule_set(cdev, &pcetable);
+	if (ret) {
+		pr_err("%s:%s:%d", __FILE__, __func__, __LINE__);
+		return GSW_statusErr;
+	}
+
+	return GSW_statusOk;
+}
+
 GSW_return_t gsw_vxlan_cfg_get(void *cdev,
 			       gsw_vxlan_cfg_t *parm)
 {
@@ -28059,6 +28131,8 @@ static void gsw_init_fn_ptrs(struct core_ops *ops)
 	ops->gsw_tflow_ops.TFLOW_PceRuleFree			= GSW_PceRuleFree;
 	ops->gsw_tflow_ops.TFLOW_GlobalRule_BitMapAlloc = GSW_GlobalRule_BitMapAlloc;
 	ops->gsw_tflow_ops.TFLOW_GlobalRule_BitMapFree  = GSW_GlobalRule_BitMapFree;
+	ops->gsw_tflow_ops.TFLOW_PceRuleEnable			= GSW_PceRuleEnable;
+	ops->gsw_tflow_ops.TFLOW_PceRuleDisable			= GSW_PceRuleDisable;
 
 	/*QOS operations*/
 	ops->gsw_qos_ops.QoS_MeterCfgGet				= GSW_QoS_MeterCfgGet;
diff --git a/drivers/net/ethernet/lantiq/switch-api/gsw_ll_func.h b/drivers/net/ethernet/lantiq/switch-api/gsw_ll_func.h
index 4494a0805bb6..0d8d4e12b07d 100644
--- a/drivers/net/ethernet/lantiq/switch-api/gsw_ll_func.h
+++ b/drivers/net/ethernet/lantiq/switch-api/gsw_ll_func.h
@@ -3295,6 +3295,9 @@ GSW_return_t GSW_DefaultMacFilterGet(void *cdev, GSW_MACFILTER_default_t *parm);
 GSW_return_t GSW_PceRuleAlloc(void *cdev, GSW_PCE_rule_alloc_t *parm);
 GSW_return_t GSW_PceRuleFree(void *cdev, GSW_PCE_rule_alloc_t *parm);
 
+GSW_return_t GSW_PceRuleEnable(void *cdev, GSW_PCE_rule_t *parm);
+GSW_return_t GSW_PceRuleDisable(void *cdev, GSW_PCE_rule_t *parm);
+
 GSW_return_t GSW_Debug_CtpTableStatus(void *cdev, GSW_debug_t *parm);
 GSW_return_t GSW_Debug_BrgPortTableStatus(void *cdev, GSW_debug_t *parm);
 GSW_return_t GSW_Debug_BrgTableStatus(void *cdev, GSW_debug_t *parm);
diff --git a/drivers/net/ethernet/lantiq/switch-api/gsw_tbl_rw.c b/drivers/net/ethernet/lantiq/switch-api/gsw_tbl_rw.c
index f2bb6f4bd252..cb5828becfc0 100644
--- a/drivers/net/ethernet/lantiq/switch-api/gsw_tbl_rw.c
+++ b/drivers/net/ethernet/lantiq/switch-api/gsw_tbl_rw.c
@@ -824,6 +824,56 @@ int gsw_pce_table_key_write(void *cdev, pctbl_prog_t *ptdata)
 	return GSW_statusOk;
 }
 
+/**
+ * gsw_pce_rule_set()- Enable/Disable PCE Rule
+ * @cdev: a pointer to the target &ethsw_api_dev_t
+ * @pctbl_prog_t: a pointer to PCE Table target &pctbl_prog_t
+ *
+ * Parse the valid bit,Table Address and set the valid bit.
+ */
+GSW_return_t gsw_pce_rule_set(void *cdev, pctbl_prog_t *ptdata)
+{
+	u32 ctrlval;
+	ethsw_api_dev_t *gswdev = GSW_PDATA_GET(cdev);
+
+	BUG_ON(!gswdev || !cdev);
+
+	do {
+		gsw_r32_raw(cdev, PCE_TBL_CTRL_BAS_OFFSET, &ctrlval);
+	} while (gsw_field_r32(ctrlval, PCE_TBL_CTRL_BAS_SHIFT,
+			       PCE_TBL_CTRL_BAS_SIZE));
+
+#ifdef __KERNEL__
+	spin_lock(&gswdev->lock_pce_tbl);
+#endif
+
+	gsw_w32_raw(cdev, PCE_TBL_ADDR_ADDR_OFFSET, ptdata->pcindex);
+	/*TABLE ADDRESS*/
+	ctrlval = 0;
+	ctrlval = gsw_field_w32(ctrlval, PCE_TBL_CTRL_ADDR_SHIFT,
+				PCE_TBL_CTRL_ADDR_SIZE, ptdata->table);
+	ctrlval = gsw_field_w32(ctrlval, PCE_TBL_CTRL_OPMOD_SHIFT,
+				PCE_TBL_CTRL_OPMOD_SIZE, PCE_OP_MODE_ADWR);
+	ctrlval = gsw_field_w32(ctrlval, PCE_TBL_CTRL_VLD_SHIFT,
+				PCE_TBL_CTRL_VLD_SIZE, ptdata->valid);
+	ctrlval = gsw_field_w32(ctrlval, PCE_TBL_CTRL_BAS_SHIFT,
+				PCE_TBL_CTRL_BAS_SIZE, 1);
+	gsw_w32_raw(cdev, PCE_TBL_CTRL_BAS_OFFSET, ctrlval);
+
+	do {
+		gsw_r32_raw(cdev, PCE_TBL_CTRL_BAS_OFFSET, &ctrlval);
+	} while (gsw_field_r32(ctrlval, PCE_TBL_CTRL_BAS_SHIFT,
+			       PCE_TBL_CTRL_BAS_SIZE));
+
+	gsw_w32_raw(cdev, PCE_TBL_CTRL_ADDR_OFFSET, 0);
+
+#ifdef __KERNEL__
+	spin_unlock(&gswdev->lock_pce_tbl);
+#endif
+
+	return GSW_statusOk;
+}
+
 
 GSW_return_t GSW_DumpTable(void *cdev, GSW_table_t *parm)
 {
diff --git a/include/net/switch_api/gsw_flow_ops.h b/include/net/switch_api/gsw_flow_ops.h
index 1a257ffb36fb..22210856a7bc 100644
--- a/include/net/switch_api/gsw_flow_ops.h
+++ b/include/net/switch_api/gsw_flow_ops.h
@@ -154,7 +154,11 @@ struct tflow_ops {
 	GSW_return_t (*TFLOW_PceRuleAlloc)(void *, GSW_PCE_rule_alloc_t *);
 	/* Command: GSW_TFLOW_FREE ; Index: 0x05 */
 	GSW_return_t (*TFLOW_PceRuleFree)(void *, GSW_PCE_rule_alloc_t *);
-	/* BitMap to Allocate usuable free Global Entries*/
+	/* Command: GSW_PCE_RULE_ENABLE ; Index: 0x06 */
+	GSW_return_t (*TFLOW_PceRuleEnable)(void *, GSW_PCE_rule_t *);
+	/* Command: GSW_PCE_RULE_DISABLE ; Index: 0x07 */
+	GSW_return_t (*TFLOW_PceRuleDisable)(void *, GSW_PCE_rule_t *);
+	/* BitMap to Allocate free Global Entries*/
 	GSW_return_t (*TFLOW_GlobalRule_BitMapAlloc)(void *, GSW_PCE_GlobalBitMap_t *);
 	/* BitMap to Free used Global Entries*/
 	GSW_return_t (*TFLOW_GlobalRule_BitMapFree)(void *, GSW_PCE_GlobalBitMap_t *);
diff --git a/include/net/switch_api/gsw_tbl_rw.h b/include/net/switch_api/gsw_tbl_rw.h
index cf91d34c2033..f7b5e0b65e28 100644
--- a/include/net/switch_api/gsw_tbl_rw.h
+++ b/include/net/switch_api/gsw_tbl_rw.h
@@ -159,6 +159,7 @@ int route_table_write(void *cdev, pctbl_prog_t *rdata);
 GSW_return_t gsw_bm_table_read(void *cdev, bmtbl_prog_t *ptdata);
 GSW_return_t gsw_bm_table_write(void *cdev, bmtbl_prog_t *ptdata);
 u32 pmac_addr_off(u32 off, u32 id);
+GSW_return_t gsw_pce_rule_set(void *cdev, pctbl_prog_t *ptdata);
 
 #endif
 
diff --git a/include/net/switch_api/lantiq_gsw_flow.h b/include/net/switch_api/lantiq_gsw_flow.h
index aeffd52b03a5..81a781e7fa0d 100644
--- a/include/net/switch_api/lantiq_gsw_flow.h
+++ b/include/net/switch_api/lantiq_gsw_flow.h
@@ -1447,6 +1447,32 @@ typedef struct {
  */
 #define GSW_PCE_RULE_FREE	_IOWR(GSW_TFLOW_MAGIC, 0x05, GSW_PCE_rule_alloc_t)
 
+/**
+ *\brief Enable PCE Rule.
+ *It is used to Enable PCE Rule written by
+ *\ref GSW_PCE_RULE_WRITE.
+
+ *\param GSW_PCE_rule_t Pointer to \ref GSW_PCE_rule_t.
+
+ *\return Return value as follows:
+ *- GSW_statusOk: if successful
+ *- An error code in case an error occurs
+ */
+#define GSW_PCE_RULE_ENABLE	_IOWR(GSW_TFLOW_MAGIC, 0x06, GSW_PCE_rule_t)
+
+/**
+ *\brief Disable PCE Rule.
+ *It is used to Disable PCE Rule written by
+ *\ref GSW_PCE_RULE_WRITE.
+
+ *\param GSW_PCE_rule_t Pointer to \ref GSW_PCE_rule_t.
+
+ *\return Return value as follows:
+ *- GSW_statusOk: if successful
+ *- An error code in case an error occurs
+ */
+#define GSW_PCE_RULE_DISABLE _IOWR(GSW_TFLOW_MAGIC, 0x07, GSW_PCE_rule_t)
+
 /*@}*/ /* GSW_IOCTL_CLASS */
 
 /** \addtogroup GSW_IOCTL_SYS */
